### PR TITLE
[ADVAPP-215]: Installation and setup of tenancy with Spatie Multi-tenancy

### PR DESCRIPTION
**Pull request by Orrison**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/ADVAPP-215

### Technical Description

Multitenancy.

### Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Content or styling update (Changes which don't affect functionality)
- [x] DevOps
- [ ] Documentation

### Any deployment steps required?

A deployment step could be a command that needs to be executed or an ENV key that needs to be added, for example.

- [x] Yes, please specify
- [ ] No

_______________________________________________

## Before contributing and submitting this PR, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/canyongbs/advisingapp/blob/main/README.md#contributing).
* [x] Title the PR with the ticket/issue number and a short description of the changes made. Or if no ticket/issue exists, title the PR with a short description of the changes made
* [x] Linked a relevant ticket or issue or describe the issue/feature which this PR resolves/implements.
* [x] Resolved all conflicts, if any.
* [x] Rebased your branch PR on top of the latest upstream `develop` branch.

### CODE REVIEW

File: app-modules/engagement/tests/EngagementFile/EngagementFilePruningTest.php [hunk 1]
```diff
@@ -69,10 +69,12 @@
 it('is scheduled to prune EngagementFiles daily during scheduler run', function () {
     $schedule = app()->make(Schedule::class);
 
+    ray($schedule->events());
```
Comment (danharrin) [inline app-modules/engagement/tests/EngagementFile/EngagementFilePruningTest.php:72]: ray

**Review by dgoetzit — CHANGES_REQUESTED** [review]
This looks great @Orrison! Just one small change, and then a few questions.

File: docs/local-setup.md [hunk 2]
```diff
@@ -64,18 +64,22 @@ sail up -d
...
-sail npm run dev
+sail npm run build
 ```
 
-Note: `npm run dev` will run vite in watch mode. If you wish to run it in production mode, you can run `npm run build` instead.
+These commands will set up the application for the "landlord" database. The landlord database is in charge of holding all information on tenants. Next we will set up a tenant.
 
-Note: It can be quicker to run `npm run dev` on your host machine if you have the correct node version installed.
+```bash
+sail artisan tenant:create [A Name for the Tenant] [A domain for the tenant]
```
[the remark this answers is not in the document]
Comment (Orrison) [reply docs/local-setup.md:75]: Yes, but we can do this in future improvements of the local setup process if that is okay with you?

File: app/Console/Commands/CreateTenantCommand.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    protected $signature = 'tenants:create {name} {domain}';
+
+    protected $description = 'Temporary command to test the tenant creation process.';
+
+    public function handle(): void
+    {
+        $name = $this->argument('name');
+        $domain = $this->argument('domain');
+
+        $database = 'tenant_' . strtolower(Str::random(30));
```
[the remark this answers is not in the document]
Comment (Orrison) [reply app/Console/Commands/CreateTenantCommand.php:62]: With 30 random chars the chances of that are so low. And this is not the code that will be used in Production. The production API endpoint will have the database information provided to it.

This is just for local development.

File: app/Console/Commands/PrintApiSchema.php [hunk 4]
```diff
@@ -34,33 +34,31 @@
...
+namespace App\Console\Commands;
 
-use Illuminate\Database\Eloquent\Model;
-use Illuminate\Database\Eloquent\Builder;
+use App\Models\Tenant;
+use Illuminate\Console\Command;
+use Illuminate\Support\Facades\Artisan;
 
-trait Tenantable
+class PrintApiSchema extends Command
```
[the remark this answers is not in the document]
Comment (Orrison) [reply app/Console/Commands/PrintApiSchema.php:43]: Not at the moment no. I have created a different ticket for making the API schema generation tenant aware as that will be move involved.

File: app/Providers/HealthServiceProvider.php [hunk 5]
```diff
@@ -64,7 +64,8 @@ public function boot(): void
             CacheCheck::new(),
             OptimizedAppCheck::new(),
             DatabaseCheck::new()
-                ->name('pgsql')
+                ->name('tenant')
+                ->connectionName('tenant')
```
Comment (mxm1070) [inline app/Providers/HealthServiceProvider.php:68]: We let them set the `tenantConnectionName` in `SwitchTenantDatabasesTask` so doesn't this need to take that into account?

File: docs/local-setup.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dgoetzit) [reply docs/local-setup.md:75]: Yep, sounds good to me.

**Review by dgoetzit — DISMISSED** [review]

**Review by mxm1070 — COMMENTED** [review]
Looks really good! Love the configs!

File: app/Console/Kernel.php [hunk 6]
```diff
@@ -56,69 +55,113 @@ class Kernel extends ConsoleKernel
...
+
+            $schedule->command("tenants:artisan \"health:check\" --tenant={$tenant->id}")
+                ->everyMinute()
+                ->onOneServer()
+                ->runInBackground();
+
+            $schedule->command("tenants:artisan \"health:queue-check-heartbeat\" --tenant={$tenant->id}")
+                ->everyMinute()
+                ->onOneServer()
+                ->runInBackground();
```
[the remark this answers is not in the document]
Comment (Orrison) [reply app/Console/Kernel.php:79]: I don't think I moved these from where they were in the order.

And I read the Spatie docs and I do not see where it says it has to be in this order. It just recommends that you put the schedule heartbeat last specifically. Which we are doing here. All commands would have been dispatched and the scheduler is working. Whether the individual processes are erroring we can asses in error logs. There also is just no way we can feasibly not run these in background.

File: app/Http/Kernel.php [hunk 7]
```diff
@@ -64,20 +81,24 @@ class Kernel extends HttpKernel
...
+            SubstituteBindings::class,
             Middleware\AuthGates::class,
             Middleware\SetPreferredLocale::class,
         ],
 
         'api' => [
-            \Laravel\Sanctum\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
+            NeedsTenant::class,
+            StartSession::class,
+            EnsureFrontendRequestsAreStateful::class,
```
[the remark this answers is not in the document]
Comment (Orrison) [reply app/Http/Kernel.php:99]: Whoops, we probably don't need this. Removing.

File: app/Http/Kernel.php [hunk 8]
```diff
@@ -50,11 +67,11 @@ class Kernel extends HttpKernel
     protected $middleware = [
         // \App\Http\Middleware\TrustHosts::class,
         Middleware\TrustProxies::class,
```
[the remark this answers is not in the document]
Comment (Orrison) [reply app/Http/Kernel.php]: Done

File: app/Providers/HealthServiceProvider.php (hunk 5 above)
Comment (Orrison) [reply app/Providers/HealthServiceProvider.php:68]: It's actually hard coded to `tenant` in the config file now. But good point, I changed to to look at the config value.

File: app/Console/Kernel.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (mxm1070) [reply app/Console/Kernel.php:79]: Ah, in the background makes sense. I didn't consider that would change the timing.

File: app/Http/Kernel.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (mxm1070) [reply app/Http/Kernel.php]: Is session also intended? I didn't think an api normally had that either.

[the remark this answers is not in the document]
Comment (Orrison) [reply app/Http/Kernel.php]: Removed

**Review by mxm1070 — APPROVED** [review]

**Pull request merged**